File: 3579.minimum-steps-to-convert-string-with-operations.py
#
# @lc app=leetcode id=3579 lang=python3
#
# [3579] Minimum Steps to Convert String with Operations
#

# @lc code=start
class Solution:
    def minOperations(self, word1: str, word2: str) -> int:
        n = len(word1)
        operations = 0
        i = 0
        while i < n:
            if word1[i] != word2[i]:
                # Find length of current mismatched segment
                j = i
                while j < n and word1[j] != word2[j]:
                    j += 1
                segment_length = j - i
                # Calculate necessary operations for this segment dynamically
                if segment_length == 1:
                    # Single character mismatch, one replace needed
                    operations += 1
                elif segment_length == 2:
                    # Optimal operation could be one swap if it aligns both characters correctly; else two replaces.
                    if (word1[i] == word2[i+1]) and (word1[i+1] == word2[i]):
                        operations += 1 # One swap fixes it perfectly
                    else:
                        operations += 2 # Two replaces in other cases
                else:
                    # Use reverse first to potentially align more characters, then handle remaining with swaps/replaces.
                    operations += (segment_length // 2) + (segment_length % 2)
                i = j # Move to next unmatched section after handling current one
            else:
                i += 1 # Move to next character if matched
        return operations 
# @lc code=end

File: 3579.minimum-steps-to-convert-string-with-operations.go
# @lc app=leetcode id=3579 lang=golang
#
# [3579] Minimum Steps to Convert String with Operations
#
# @lc code=start
func minOperations(word1 string, word2 string) int {
    n := len(word1)
    m := len(word2)
    
    // Initialize a DP table for storing minimum operations
    dp := make([][]int, n+1)
    for i := range dp {
        dp[i] = make([]int, m+1)
    }

    // Base cases: converting empty prefix to another requires edits equal to length difference
    for i := 0; i <= n; i++ {
        dp[i][0] = i
    }
    for j := 0; j <= m; j++ {
        dp[0][j] = j
    }

    // Fill the DP table considering all operations
    for i := 1; i <= n; i++ {
        for j := 1; j <= m; j++ {
            if word1[i-1] == word2[j-1] {
                dp[i][j] = dp[i-1][j-1]
            } else {
                // Calculate costs for replace, swap, and reverse where applicable
                replaceCost := dp[i-1][j-1] + 1
                minCost := replaceCost
                if i > 1 && j > 1 && word1[i-2] == word2[j-1] && word2[j-2] == word1[i-1] {
                    swapCost := dp[i-2][j-2] + 1 
                    minCost = min(minCost, swapCost)
                }
                
dp[i][j] = min(minCost, dp[i-1][j]+1, dp[i][j-1]+1)
dp[n][m]	}	}	}	return	func	min(a,	b,	c	int	int	{	if	a	<	b	&&	a	<	c	return	a	else	if	b	<	c	return	b	else	return	c# @lc code=end